File: ex11/Atividade.java
import java.util.ArrayList;
import java.util.List;
import java.util.HashMap;
import java.util.Map;


interface EstadoRobo {

    int atacar(Robo robo);

    void receberDano(Robo robo, int pontos);

    void reparar(Robo robo, int pontos);

    String nome();
}


class EstadoNormal implements EstadoRobo {

    public int atacar(Robo robo) {
        // TODO
        throw new UnsupportedOperationException();
    }

    public void receberDano(Robo robo, int pontos) {
        // TODO
        throw new UnsupportedOperationException();
    }

    public void reparar(Robo robo, int pontos) {
        // TODO
        throw new UnsupportedOperationException();
    }

    public String nome() {
        return "NORMAL";
    }
}


class EstadoCritico implements EstadoRobo {

    public int atacar(Robo robo) {
        // TODO
        throw new UnsupportedOperationException();
    }

    public void receberDano(Robo robo, int pontos) {
        // TODO
        throw new UnsupportedOperationException();
    }

    public void reparar(Robo robo, int pontos) {
        // TODO
        throw new UnsupportedOperationException();
    }

    public String nome() {
        return "CRITICO";
    }
}


class EstadoDesativado implements EstadoRobo {

    public int atacar(Robo robo) {
        // TODO
        throw new UnsupportedOperationException();
    }

    public void receberDano(Robo robo, int pontos) {
        // TODO
        throw new UnsupportedOperationException();
    }

    public void reparar(Robo robo, int pontos) {
        // TODO
        throw new UnsupportedOperationException();
    }

    public String nome() {
        return "DESATIVADO";
    }
}


class Robo {

    private String modelo;
    private int vidaMaxima;
    private int vidaAtual;
    private int ataqueBase;
    private List<String> modulos;
    private EstadoRobo estado;

    public Robo(
            String modelo,
            int vidaMaxima,
            int ataqueBase,
            List<String> modulos
    ) {
        this.modelo = modelo;
        this.vidaMaxima = vidaMaxima;
        this.vidaAtual = vidaMaxima;
        this.ataqueBase = ataqueBase;
        this.modulos = new ArrayList<>(modulos);
        this.estado = new EstadoNormal();
    }

    public int atacar() {
        // TODO: delegar ao estado atual
        throw new UnsupportedOperationException();
    }

    public void receberDano(int pontos) {
        // TODO: delegar ao estado atual
        throw new UnsupportedOperationException();
    }

    public void reparar(int pontos) {
        // TODO: delegar ao estado atual
        throw new UnsupportedOperationException();
    }

    public Robo clonar() {
        // TODO: implementar Prototype
        throw new UnsupportedOperationException();
    }

    public void mudarEstado(EstadoRobo novoEstado) {
        this.estado = novoEstado;
    }

    void reduzirVida(int pontos) {
        if (pontos > 0) {
            vidaAtual = Math.max(0, vidaAtual - pontos);
        }
    }

    void aumentarVida(int pontos) {
        if (pontos > 0) {
            vidaAtual = Math.min(
                    vidaMaxima,
                    vidaAtual + pontos
            );
        }
    }

    public void adicionarModulo(String modulo) {
        modulos.add(modulo);
    }

    public String getModelo() {
        return modelo;
    }

    public int getVidaMaxima() {
        return vidaMaxima;
    }

    public int getVidaAtual() {
        return vidaAtual;
    }

    public int getAtaqueBase() {
        return ataqueBase;
    }

    public List<String> getModulos() {
        return modulos;
    }

    public String getEstado() {
        return estado.nome();
    }
}


class CatalogoRobos {

    private Map<String, Robo> prototipos = new HashMap<>();

    public void registrar(String nome, Robo prototipo) {
        prototipos.put(nome, prototipo);
    }

    public Robo criar(String nome) {
        // TODO: localizar o prototipo e clona-lo
        throw new UnsupportedOperationException();
    }
}


public class Atividade {
    // Nao e necessario implementar main.
}

File: ex11/atividade.py
from abc import ABC, abstractmethod


class EstadoRobo(ABC):

    @abstractmethod
    def atacar(self, robo):
        pass

    @abstractmethod
    def receber_dano(self, robo, pontos):
        pass

    @abstractmethod
    def reparar(self, robo, pontos):
        pass

    @abstractmethod
    def nome(self):
        pass


class EstadoNormal(EstadoRobo):

    def atacar(self, robo):
        # TODO
        raise NotImplementedError

    def receber_dano(self, robo, pontos):
        # TODO
        raise NotImplementedError

    def reparar(self, robo, pontos):
        # TODO
        raise NotImplementedError

    def nome(self):
        return "NORMAL"


class EstadoCritico(EstadoRobo):

    def atacar(self, robo):
        # TODO
        raise NotImplementedError

    def receber_dano(self, robo, pontos):
        # TODO
        raise NotImplementedError

    def reparar(self, robo, pontos):
        # TODO
        raise NotImplementedError

    def nome(self):
        return "CRITICO"


class EstadoDesativado(EstadoRobo):

    def atacar(self, robo):
        # TODO
        raise NotImplementedError

    def receber_dano(self, robo, pontos):
        # TODO
        raise NotImplementedError

    def reparar(self, robo, pontos):
        # TODO
        raise NotImplementedError

    def nome(self):
        return "DESATIVADO"


class Robo:

    def __init__(self, modelo, vida_maxima, ataque_base, modulos=None):
        self._modelo = modelo
        self._vida_maxima = vida_maxima
        self._vida_atual = vida_maxima
        self._ataque_base = ataque_base
        self._modulos = list(modulos) if modulos is not None else []
        self._estado = EstadoNormal()

    def atacar(self):
        # TODO: delegar ao estado atual
        raise NotImplementedError

    def receber_dano(self, pontos):
        # TODO: delegar ao estado atual
        raise NotImplementedError

    def reparar(self, pontos):
        # TODO: delegar ao estado atual
        raise NotImplementedError

    def clonar(self):
        # TODO: implementar Prototype
        raise NotImplementedError

    def mudar_estado(self, novo_estado):
        self._estado = novo_estado

    def _reduzir_vida(self, pontos):
        if pontos > 0:
            self._vida_atual = max(0, self._vida_atual - pontos)

    def _aumentar_vida(self, pontos):
        if pontos > 0:
            self._vida_atual = min(
                self._vida_maxima,
                self._vida_atual + pontos
            )

    def adicionar_modulo(self, modulo):
        self._modulos.append(modulo)

    def get_modelo(self):
        return self._modelo

    def get_vida_maxima(self):
        return self._vida_maxima

    def get_vida_atual(self):
        return self._vida_atual

    def get_ataque_base(self):
        return self._ataque_base

    def get_modulos(self):
        return self._modulos

    def get_estado(self):
        return self._estado.nome()


class CatalogoRobos:

    def __init__(self):
        self._prototipos = {}

    def registrar(self, nome, prototipo):
        self._prototipos[nome] = prototipo

    def criar(self, nome):
        # TODO: localizar o prototipo e clona-lo
        raise NotImplementedError